frontend/catalyst/jax_tracer.py

```diff
@@ -541,47 +675,72 @@ def trace_quantum_function(
             def is_leaf(obj):
                 return isinstance(obj, qml.measurements.MeasurementProcess)
 
             # 2. Create a new tree that has measurements as leaves
-            ans, out_tree = jax.tree_util.tree_flatten(ans, is_leaf=is_leaf)
-
-            out_classical_tracers_or_measurements = [
-                (trace.full_raise(t) if isinstance(t, DynamicJaxprTracer) else t) for t in ans
-            ]
-
-        # (2) - Quantum tracing
-        with EvaluationContext.frame_tracing_context(ctx, trace):
-            qdevice_p.bind(spec="kwargs", val=str(device.backend_kwargs))
-            qdevice_p.bind(spec="backend", val=device.backend_name)
-            qreg_in = qalloc_p.bind(len(device.wires))
-            qrp_out = trace_quantum_tape(quantum_tape, device, qreg_in, ctx, trace)
-            out_classical_tracers, out_classical_tree = trace_quantum_measurements(
-                device,
-                qrp_out,
-                out_classical_tracers_or_measurements,
-                out_tree,
-            )
-            out_quantum_tracers = [qrp_out.actualize()]
-            qdealloc_p.bind(qreg_in)
-
-            out_classical_tracers = [trace.full_raise(t) for t in out_classical_tracers]
-
-            jaxpr, out_type, consts = ctx.frames[trace].to_jaxpr2(
-                out_classical_tracers + out_quantum_tracers
+            return_values_flat, return_values_tree = jax.tree_util.tree_flatten(
+                return_values, is_leaf=is_leaf
             )
-            jaxpr._outvars = jaxpr._outvars[:-1]  # pylint: disable=protected-access
-            out_type = out_type[:-1]
-            # TODO: `check_jaxpr` complains about the `AbstractQreg` type. Consider fixing.
-            # check_jaxpr(jaxpr)
 
-    closed_jaxpr = ClosedJaxpr(jaxpr, consts)
-    out_avals, _ = unzip2(out_type)
+            # TODO: In order to compose transforms, we would need to recursively
+            # call apply_transform while popping the latest transform applied,
+            # until there are no more transforms to be applied.
+            # But first we should clean this up this method a bit more.
+            tapes, post_processing = apply_transform(qnode, quantum_tape, return_values_flat)
 
-    abstract_results = tree_unflatten(
-        out_classical_tree, [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals]
-    )
-    return closed_jaxpr, abstract_results
+        # (2) - Quantum tracing
+        results_tracers, results_abstract = [], []
+        is_program_transformed = qnode and qnode.transform_program
+        for tape in tapes:
+            # If the program is batched, that means that it was transformed.
+            # If it was transformed, that means that the program might have
+            # changed the output. See `split_non_commuting`
+            if is_program_transformed:
+                # TODO: In the future support arbitrary output from the user function.
+                output = tape.measurements
```

Comment:
Agreed that it would be difficult to support this right now, since the QNode tapes don't track classical values.. However, is there a way we could raise an error if the user uses a transform and returns classical values?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/280/commits/0ef5d9a9d53e568fdc172b4bddc00dfff67f9743